src/sched/lthread.c

```diff
@@ -494,10 +493,23 @@ int _lthread_resume(struct lthread* lt)
 
     set_tls_tp(lt);
     _switch(&lt->ctx, &sched->ctx);
+    // The first "startmain" thread eventually loads the app's ELF image
+    // and initializes its tls area. As lthread has to properly set the
+    // tls region on context switches, check if the fs has changed and
+    // update the lthread's thread pointer field accordingly.
+    if (sched->current_lthread->tid == 1 &&
+        (sched->current_lthread)->attr.thread_type == LKL_KERNEL_THREAD){
+        void* fs_ptr;
+        __asm__ __volatile__("mov %%fs:0,%0" : "=r"(fs_ptr));
+        if (fs_ptr != sched->current_lthread->tp){
+            sched->current_lthread->tp = fs_ptr;
```

Comment:
Isn't this initialized during lthread_create (with the init_tp call)?

Reply:
`init_tp` calculates the thread pointer value and sets a self pointer, but it does not set fsbase. FS register is set during context switches in `set_tls_tp`.

This above code is required because for the `startmain` thread, the dynamic loader creates a new tls region before passing execution to it - [here](https://github.com/vtikoo/sgx-lkl-musl/blob/move-lthread-out/ldso/dynlink.c#L1951-L1959).
lthreads needs to save that new pointer.